### PR TITLE
fix(pkg): correctly verify tarball checksums

### PR DESCRIPTION
**Pull request by rgrinberg**
The changes I've made to the opam libraries are rather ugly, but it does seem like their api needs to be redesigned a bit to allow more customized error handling. This isn't a priority for now though.

### CODE REVIEW

**Comment by Leonidas-from-XIV** [comment]
What is the fix here? When I wrote the original code I tried to make sure to avoid having to add additional patches to our OPAM fork, because the way it is currently computed in OPAM is very much unusable.

**Comment by Alizter** [comment]
@Leonidas-from-XIV Now that we unpack downloads the file being checksummed is a directory which breaks the opam code.

File: test/expect-tests/dune_pkg/fetch_tests.ml [hunk 1]
```diff
@@ -186,5 +191,8 @@ let%expect_test "downloading, tarball" =
   (Dune_util__Report_error.Already_reported)
   Trailing output
   ---------------
-  Error: Is a directory |}]
+  Error: Expected checksum was
+  md5=c533195dc4253503071a19d42f08e877
+  but got
+  <REDACTED> |}]
```
Comment (Alizter) [inline test/expect-tests/dune_pkg/fetch_tests.ml:197]: @Leonidas-from-XIV  See the test here.

**Comment by Leonidas-from-XIV** [comment]
Ah, but then shouldn't the OPAM patches be applied to our fork first and then updated here? Otherwise the one who updates it next will accidentally undo these patches.

**Comment by Leonidas-from-XIV** [comment]
The problem is that this changes the API so whatever works here might need additional changes in our OPAM fork to continue building (e.g. handling of the new variant).

**Review by Leonidas-from-XIV — COMMENTED** [review]
I think it would be nice to have something that we could eventually upstream (shouldn't be too controversial) so we don't have to maintain a patch like this.

**Comment by rgrinberg** [comment]
Yes, it would be nice to have something that we can upstream. For now, I’m afraid we’ll have to work with this hot fix. Without this, the entire feature is not usable in practice so it’s a bit urgent to get something working.

**Comment by Alizter** [comment]
@rgrinberg Were you going to update the fork?

**Pull request merged**